File: src/main/java/no/nav/syfo/service/MotebehovService.java
package no.nav.syfo.service;

import no.nav.syfo.aktorregister.AktorregisterConsumer;
import no.nav.syfo.aktorregister.domain.Fodselsnummer;
import no.nav.syfo.behandlendeenhet.BehandlendeEnhetConsumer;
import no.nav.syfo.domain.rest.*;
import no.nav.syfo.exception.ConflictException;
import no.nav.syfo.repository.dao.MotebehovDAO;
import no.nav.syfo.repository.domain.PMotebehov;
import no.nav.syfo.util.Metrikk;
import org.slf4j.Logger;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import javax.inject.Inject;
import java.util.*;

import static java.util.Collections.emptyList;
import static java.util.stream.Collectors.toList;
import static org.slf4j.LoggerFactory.getLogger;

@Service
public class MotebehovService {

    private static final Logger log = getLogger(MotebehovService.class);

    private final Metrikk metrikk;
    private final AktorregisterConsumer aktorregisterConsumer;
    private final BehandlendeEnhetConsumer behandlendeEnhetConsumer;
    private final OversikthendelseService oversikthendelseService;
    private final MotebehovDAO motebehovDAO;

    @Inject
    public MotebehovService(
            Metrikk metrikk,
            AktorregisterConsumer aktorregisterConsumer,
            BehandlendeEnhetConsumer behandlendeEnhetConsumer,
            OversikthendelseService oversikthendelseService,
            MotebehovDAO motebehovDAO
    ) {
        this.metrikk = metrikk;
        this.aktorregisterConsumer = aktorregisterConsumer;
        this.behandlendeEnhetConsumer = behandlendeEnhetConsumer;
        this.oversikthendelseService = oversikthendelseService;
        this.motebehovDAO = motebehovDAO;
    }

    @Transactional
    public void behandleUbehandledeMotebehov(final Fodselsnummer arbeidstakerFnr, final String veilederIdent) {
        int antallOppdateringer = motebehovDAO.oppdaterUbehandledeMotebehovTilBehandlet(aktorregisterConsumer.getAktorIdForFodselsnummer(arbeidstakerFnr), veilederIdent);

        if (antallOppdateringer > 0) {
            String behandlendeEnhet = behandlendeEnhetConsumer.getBehandlendeEnhet(arbeidstakerFnr.getValue(), null).getEnhetId();
            oversikthendelseService.sendOversikthendelse(arbeidstakerFnr.getValue(), behandlendeEnhet);
        } else {
            metrikk.tellHendelse("feil_behandle_motebehov_svar_eksiterer_ikke");
            log.warn("Ugyldig tilstand: Veileder {} forsøkte å behandle motebehovsvar som ikke eksisterer. Kaster Http-409", veilederIdent);
            throw new ConflictException();
        }
    }

    public List<Motebehov> hentMotebehovListe(final Fodselsnummer arbeidstakerFnr) {
        final String arbeidstakerAktoerId = aktorregisterConsumer.getAktorIdForFodselsnummer(arbeidstakerFnr);
        return motebehovDAO.hentMotebehovListeForAktoer(arbeidstakerAktoerId)
                .orElse(emptyList())
                .stream()
                .map(dbMotebehov -> mapPMotebehovToMotebehov(arbeidstakerFnr, dbMotebehov))
                .collect(toList());
    }

    public List<Motebehov> hentMotebehovListeForOgOpprettetAvArbeidstaker(final Fodselsnummer arbeidstakerFnr) {
        final String arbeidstakerAktoerId = aktorregisterConsumer.getAktorIdForFodselsnummer(new Fodselsnummer(arbeidstakerFnr.getValue()));
        return motebehovDAO.hentMotebehovListeForOgOpprettetAvArbeidstaker(arbeidstakerAktoerId)
                .orElse(emptyList())
                .stream()
                .map(dbMotebehov -> mapPMotebehovToMotebehov(arbeidstakerFnr, dbMotebehov))
                .collect(toList());
    }

    public List<Motebehov> hentMotebehovListeForArbeidstakerOpprettetAvLeder(final Fodselsnummer arbeidstakerFnr, String virksomhetsnummer) {
        final String arbeidstakerAktoerId = aktorregisterConsumer.getAktorIdForFodselsnummer(new Fodselsnummer(arbeidstakerFnr.getValue()));
        return motebehovDAO.hentMotebehovListeForArbeidstakerOpprettetAvLeder(arbeidstakerAktoerId, virksomhetsnummer)
                .orElse(emptyList())
                .stream()
                .map(dbMotebehov -> mapPMotebehovToMotebehov(arbeidstakerFnr, dbMotebehov))
                .collect(toList());
    }

    @Transactional
    public UUID lagreMotebehov(Fodselsnummer innloggetFNR, Fodselsnummer arbeidstakerFnr, final NyttMotebehov nyttMotebehov) {
        final String innloggetBrukerAktoerId = aktorregisterConsumer.getAktorIdForFodselsnummer(innloggetFNR);
        final String arbeidstakerAktoerId = aktorregisterConsumer.getAktorIdForFodselsnummer(arbeidstakerFnr);
        final String arbeidstakerBehandlendeEnhet = behandlendeEnhetConsumer.getBehandlendeEnhet(arbeidstakerFnr.getValue(), null).getEnhetId();
        final PMotebehov motebehov = mapNyttMotebehovToPMotebehov(innloggetBrukerAktoerId, arbeidstakerAktoerId, arbeidstakerBehandlendeEnhet, nyttMotebehov);

        UUID id = motebehovDAO.create(motebehov);

        if (nyttMotebehov.motebehovSvar.harMotebehov) {
            oversikthendelseService.sendOversikthendelse(nyttMotebehov
                    .arbeidstakerFnr(arbeidstakerFnr.getValue())
                    .tildeltEnhet(arbeidstakerBehandlendeEnhet)
            );
        }
        return id;
    }

    private PMotebehov mapNyttMotebehovToPMotebehov(String innloggetAktoerId, String arbeidstakerAktoerId, String tildeltEnhet, NyttMotebehov nyttMotebehov) {
        return new PMotebehov()
                .opprettetAv(innloggetAktoerId)
                .aktoerId(arbeidstakerAktoerId)
                .virksomhetsnummer(nyttMotebehov.virksomhetsnummer)
                .harMotebehov(nyttMotebehov.motebehovSvar().harMotebehov)
                .forklaring(nyttMotebehov.motebehovSvar().forklaring)
                .tildeltEnhet(tildeltEnhet)
                .behandletTidspunkt(null);
    }

    private Motebehov mapPMotebehovToMotebehov(Fodselsnummer arbeidstakerFnr, PMotebehov pMotebehov) {
        return new Motebehov()
                .id(pMotebehov.uuid)
                .opprettetDato(pMotebehov.opprettetDato)
                .aktorId(pMotebehov.aktoerId)
                .opprettetAv(pMotebehov.opprettetAv)
                .arbeidstakerFnr(arbeidstakerFnr.getValue())
                .virksomhetsnummer(pMotebehov.virksomhetsnummer)
                .motebehovSvar(new MotebehovSvar()
                        .harMotebehov(pMotebehov.harMotebehov)
                        .forklaring(pMotebehov.forklaring)
                )
                .tildeltEnhet(pMotebehov.tildeltEnhet)
                .behandletTidspunkt(pMotebehov.behandletTidspunkt)
                .behandletVeilederIdent(pMotebehov.behandletVeilederIdent);
    }
}

File: src/main/kotlin/no/nav/syfo/motebehov/MotebehovService.kt
package no.nav.syfo.motebehov

import no.nav.syfo.api.exception.ConflictException
import no.nav.syfo.consumer.behandlendeenhet.BehandlendeEnhetConsumer
import no.nav.syfo.consumer.behandlendeenhet.getEnhetId
import no.nav.syfo.consumer.pdl.PdlConsumer
import no.nav.syfo.metric.Metric
import no.nav.syfo.motebehov.database.MotebehovDAO
import no.nav.syfo.motebehov.database.toMotebehov
import no.nav.syfo.motebehov.motebehovstatus.MotebehovSkjemaType
import no.nav.syfo.personoppgavehendelse.PersonoppgavehendelseService
import org.slf4j.LoggerFactory
import org.springframework.dao.DataAccessException
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import java.time.LocalDate
import java.time.LocalDateTime
import java.util.*
import java.util.stream.Collectors
import javax.inject.Inject

@Service
class MotebehovService @Inject constructor(
    private val metric: Metric,
    private val behandlendeEnhetConsumer: BehandlendeEnhetConsumer,
    private val personoppgavehendelseService: PersonoppgavehendelseService,
    private val motebehovDAO: MotebehovDAO,
    private val pdlConsumer: PdlConsumer,
) {
    @Transactional
    fun behandleUbehandledeMotebehov(arbeidstakerFnr: String, veilederIdent: String) {
        val arbeidstakerAktorId = pdlConsumer.aktorid(arbeidstakerFnr)
        val pMotebehovUbehandletList = motebehovDAO.hentUbehandledeMotebehov(arbeidstakerAktorId)
        if (pMotebehovUbehandletList.isNotEmpty()) {
            pMotebehovUbehandletList.forEach { pMotebehov ->
                val antallOppdatering =
                    motebehovDAO.oppdaterUbehandledeMotebehovTilBehandlet(pMotebehov.uuid, veilederIdent)
                if (antallOppdatering == 1) {
                    personoppgavehendelseService.sendPersonoppgaveHendelseBehandlet(pMotebehov.uuid, arbeidstakerFnr)
                }
            }
        } else {
            metric.tellHendelse("feil_behandle_motebehov_svar_eksiterer_ikke")
            log.warn(
                "Ugyldig tilstand: Veileder {} forsøkte å behandle motebehovsvar som ikke eksisterer. Kaster Http-409",
                veilederIdent,
            )
            throw ConflictException()
        }
    }

    @Transactional
    fun behandleUbehandledeMotebehovOpprettetTidligereEnnDato(dato: LocalDate, veilederIdent: String): Int {
        var updatedCount = 0
        motebehovDAO.hentUbehandledeMotebehovEldreEnnDato(dato)
            .forEach { pMotebehov ->
                pMotebehov.sykmeldtFnr?.let {
                    val antallOppdatering =
                        motebehovDAO.oppdaterUbehandledeMotebehovTilBehandlet(pMotebehov.uuid, veilederIdent)
                    if (antallOppdatering == 1) {
                        personoppgavehendelseService.sendPersonoppgaveHendelseBehandlet(pMotebehov.uuid, it)
                        updatedCount++
                    }
                }
            }
        return updatedCount
    }

    fun hentMotebehovListe(arbeidstakerFnr: String): List<Motebehov> {
        val arbeidstakerAktoerId = pdlConsumer.aktorid(arbeidstakerFnr)
        return motebehovDAO.hentMotebehovListeForAktoer(arbeidstakerAktoerId)
            .stream()
            .map { it.toMotebehov(arbeidstakerFnr) }
            .collect(Collectors.toList())
    }

    fun hentMotebehov(motebehovId: String): Motebehov? {
        return motebehovDAO
            .hentMotebehov(motebehovId)
            .stream()
            .map { it.toMotebehov() }
            .collect(Collectors.toList())
            .firstOrNull()
    }

    fun hentMotebehovListeForOgOpprettetAvArbeidstaker(arbeidstakerFnr: String): List<Motebehov> {
        val arbeidstakerAktoerId = pdlConsumer.aktorid(arbeidstakerFnr)
        return motebehovDAO.hentMotebehovListeForOgOpprettetAvArbeidstaker(arbeidstakerAktoerId)
            .stream()
            .map { it.toMotebehov(arbeidstakerFnr) }
            .collect(Collectors.toList())
    }

    fun hentMotebehovListeForArbeidstakerOpprettetAvLeder(
        arbeidstakerFnr: String,
        isOwnLeader: Boolean,
        virksomhetsnummer: String,
    ): List<Motebehov> {
        val arbeidstakerAktoerId = pdlConsumer.aktorid(arbeidstakerFnr)
        return motebehovDAO.hentMotebehovListeForArbeidstakerOpprettetAvLeder(
            arbeidstakerAktoerId,
            isOwnLeader,
            virksomhetsnummer,
        )
            .stream()
            .map { it.toMotebehov(arbeidstakerFnr) }
            .collect(Collectors.toList())
    }

    @Transactional
    fun lagreMotebehov(
        innloggetFNR: String,
        arbeidstakerFnr: String,
        virksomhetsnummer: String,
        skjemaType: MotebehovSkjemaType,
        innmelderType: MotebehovInnmelderType,
        motebehovFormSubmission: MotebehovFormSubmissionDTO,
    ): UUID {
        val innloggetBrukerAktoerId = pdlConsumer.aktorid(innloggetFNR)
        val arbeidstakerAktoerId = pdlConsumer.aktorid(arbeidstakerFnr)
        val arbeidstakerBehandlendeEnhet =
            behandlendeEnhetConsumer.getBehandlendeEnhet(arbeidstakerFnr, null).getEnhetId()

        val motebehov = Motebehov(
            id = UUID.randomUUID(),
            opprettetDato = LocalDateTime.now(),
            aktorId = arbeidstakerAktoerId,
            opprettetAv = innloggetBrukerAktoerId,
            opprettetAvFnr = innloggetFNR,
            arbeidstakerFnr = arbeidstakerFnr,
            virksomhetsnummer = virksomhetsnummer,
            formSubmission = motebehovFormSubmission,
            tildeltEnhet = arbeidstakerBehandlendeEnhet,
            skjemaType = skjemaType,
            innmelderType = innmelderType,
        )

        val pMotebehov = motebehov.toPMotebehov()

        val uuid: UUID

        try {
            uuid = motebehovDAO.create(pMotebehov)
        } catch (ex: DataAccessException) {
            metric.tellHendelse("feil_lagre_motebehov")

            log.error(
                "DataAccessException ved lagring av motebehov med skjemaType {} og innmelderType {}: {}",
                skjemaType.name,
                innmelderType.name,
                ex.message,
            )

            throw ex
        } catch (ex: IllegalStateException) {
            metric.tellHendelse("feil_lagre_motebehov")

            log.error(
                "IllegalStateException ved lagring av motebehov med skjemaType {} og innmelderType {}: {}",
                skjemaType.name,
                innmelderType.name,
                ex.message,
            )

            throw ex
        }

        if (motebehovFormSubmission.harMotebehov) {
            personoppgavehendelseService.sendPersonoppgaveHendelseMottatt(uuid, arbeidstakerFnr)
        }
        return uuid
    }

    companion object {
        private val log = LoggerFactory.getLogger(MotebehovService::class.java)
    }
}
